### ISSUE TITLE
Excessive memory use by "gobgp mrt inject global"

### ISSUE THREAD

**Issue by vincentbernat** [osrg/gobgp#698]
Hi!

I am trying to inject a full view on a machine with limited memory using `gobgp mrt inject global latest-bview`. gobgp is being killed by oom-killer:

```
[   53.295983] Out of memory: Kill process 203 (gobgp) score 605 or sacrifice child
[   53.296494] Killed process 203 (gobgp) total-vm:1396212kB, anon-rss:1279452kB, file-rss:0kB
```

gobgpd is still here and using its share of memory:

```
root       192  6.8 35.7 1067212 733148 ?      Sl   17:28   0:17 /mnt/lab/gobgpd -f /mnt/lab/gobgp.U2.conf -t yaml -p -s unix --disable-stdlog
```

gobgp shouldn't need much memory as it should only be inject routes to gobgpd.

**Comment by fujita** [comment]
Hmm, the log says that gobgp (client command) died due to OOM. gobgpd (bgp speaker daemon) is continues to work, as expected.

"gobgp mrt inject" tries to read lots of entries from a file and concurrently sends them to gobgpd via the RPC API. So gobgp could use lots of memory temporarily.

We focus on scalability and performance rather than memory usage.

**Comment by fujita** [comment]
I disabled grpc trace feature. Also seems that there were memory leak bugs in grpc library. Please try the latest code (don't forget to update all the dependent libraries). gobgp (CLI) uses about 1g on my environment at maximum when try to inject full routes. Please reopen if you hit the same issue.

**Comment by XANi** [comment]
I still see that with 1.24 and current master

Trying to insert global route dump from RIPE bumps gobgp to 14 GB of RAM usage and then it fails with error (I have 32GB of ram so it doesn't hit OOM):

    -> ᛯ ./gobgp mrt inject global bview.20170601.1600                                                                 1
    failed to send: EOF

after failure it imported only ~36k routes

**Comment by XANi** [comment]
After adding some adhoc debugging (the original message is not helpful in any way), it appears to fail in main loop on internal gRPC EOF error.

I've replaced any `if err == io.EOF {return err}` in grpc lib with `fmt.Errorf("gRPC err: %s", err)` and added some assorted debugging and that's the result:

    failed to send, error during sending path [[{ 2001:252::/32 | src: local, nh: 2001:7f8:54:5::13 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::10 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::123 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::136 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::10 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::177 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::10 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::10 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::1:46 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::10 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::228 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::246 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::36 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::61 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::10 } { 2001:252::/32 | src: local, nh: 2001:7f8:54::74 }]]: gRPC err in SendMsg: EOF

**Comment by XANi** [comment]
Also I've noticed that bounding channel to something small like 128 still keeps gobgpd busy without exploding memory usage so that huge channel size limit is unnecessary. trying 128->1280 didn't really show any improvement too.

Gobgpd is stuck at around 1-2 core usage most of the time (there are spikes but I'm guessing its GC) and gobgp is below that

I've managed to make it work with big route file just by bruteforce retry ( https://github.com/efigence/gobgp/tree/crappy-fix-mtr-send ) but obviously that's just a workaround.
